build_tools/cmake/iree_filter_test.cmake

```diff
@@ -0,0 +1,65 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+include(CMakeParseArguments)
+
+function(iree_filter_test)
+  cmake_parse_arguments(
+    _FUNC
+    ""
+    "RESULT_VAR_ENABLED;SIZE;DRIVER"
+    "LABELS"
+    ${ARGN}
+  )
+  # Default to excluding; then exclusion cases are implemented as early-returns;
+  # then at the end we will override that to TRUE.
+  set("${_FUNC_RESULT_VAR_ENABLED}" FALSE PARENT_SCOPE)
```

Comment:
Interesting idea, but I'm finding this difficult to read, since each "return()" really means "disable the test, then return".

A few ideas:

* Add a macro that sets the variable and returns and call that in each case
* Set the https://cmake.org/cmake/help/latest/prop_test/DISABLED.html property instead of `if(NOT _ENABLED) return() endif()`. That might not work in all cases though, since the return in files like `iree_check_test` avoids touching configuration-specific variables. Could leave that for a second change, so this one is more of an NFC

Reply:
The property idea is interesting, but it's not clear that it would result in skipping all the burden that we want to skip. For example, in `iree_check_test`, we want to skip the `iree_bytecode_module` call altogether, so that we also avoid the `build iree-test-deps` overhead, not just the runtime `ctest` overhead.

This function is really just a chain of early-returns before we finally enable the test at the end. The only twist is that since CMake functions don't have a return value and we have to write to a parent-scope variable, we have to start by setting it to false. I don't see a better way to write this code, feel free if you see one, but I am not sure what would be gained by a macro wrapping the return.

Reply:
This is what I had in mind for a macro:

```cmake
macro(iree_disable_test_and_return) {
  set("${_FUNC_RESULT_VAR_ENABLED}" FALSE PARENT_SCOPE)
  return()
}

function(iree_filter_test) {
  ...
  # Architecture-based exclusions.
  if ((IREE_ARCH MATCHES "^riscv_") AND ("noriscv" IN_LIST _FUNC_LABELS))
    iree_disable_test_and_return()
  endif()

  ...
  set("${_FUNC_RESULT_VAR_ENABLED}" TRUE PARENT_SCOPE)
}
```

Alternately... why return at all? You could just start an "ENABLED" variable as true, flip it to false in any of the branches, then finally set it in the parent scope at the end of the function.

Reply:
> Alternately... why return at all? You could just start an "ENABLED" variable as true, flip it to false in any of the branches, then finally set it in the parent scope at the end of the function.

oh... of course. why didn't I think of that? done.